### ISSUE TITLE
[Feature Request] Recurrent DQN Example

### ISSUE THREAD

**Issue by smorad** [pytorch/rl#849]
## Motivation
As far as I can see, there is no example for recurrent DQN. I'm currently digging through the codebase trying to figure this out, but it would be nice to have an example. It's not clear where I need to make changes (e.g. collectors, replay buffer, loss function, etc). This would also be useful for implementing recurrent SAC, etc.

## Solution

A recurrent DQN/DDPG/SAC tutorial. Ideally, the recurrent portion of the model would be split out from the critic/actor.

E.g.
```python
encoder = TensorDictModule(nn.LazyLinear(32), in_keys=["observation"], out_keys=["hidden"])
memory = TensorDictModule(nn.LSTM(32, 32), in_keys=["hidden", "recurrent_state"], out_keys=["markov_state", "recurrent_state"])
qnet = DuelingMLPQNet(...)
actor = QValueActor(qnet, in_keys=["markov_state"], ...)
actor_explore = EGreedyWrapper(actor, ...)
# Can we just have the loss function work off of the markov_state
# instead of rerunning any recurrent computation?
loss_fn = DQNLoss(qnet)

# What is the correct way to use a collector here?
# should we pass in:
# policy=TensorDictSequential([encoder, memory, qnet])?
collector = MultiaSyncDataCollector(...)

# I assume special stuff will need to be done in the replay buffer
# to not shuffle transitions.
data = buffer.sample() 

encoder(data)
# Assign the markov_state key
memory(data)
# Now the rest is traditional (non-recurrent) DQN
actor(data)
# No need to use a different loss function at this point
loss = loss_fn(data)
loss.backward()
opt.step()
# Can we just call collector.update_policy_weights_()?
# Does it know how to do this correctly if we didn't use the `TensorDictSequential` here
# that was used in Collector(policy=TensorDictSequential...)?
  
```

## Alternatives

Give me some tips ;)

## Checklist

- [x] I have checked that there is no similar issue in the repo (**required**)

**Comment by vmoens** [comment]
I can give it a shot! I'll ping you when it's done

**Comment by smorad** [comment]
If you haven't already started this, I can give it a go!

**Comment by vmoens** [comment]
Sure sorry i've been caught by many other things.
If you want to work on this, you can create a tutorial inspired by the DQN tutorial maybe? And add the RNN to it. We have an LSTM module that should work with TensorDictModule, envs and datacollectors in `torchrl.modules` I believe.
Don't hesitate to open a draft PR soon, even if you just have a few lines of code, that will help us talk about it

**Comment by smorad** [comment]
@vmoens I'm running into some issues with how trajectories are handled. In a for loop, the collector seems to assign trajectory_ids starting from 0:

```python
collector = Collector(split_trajs=False)
for data in collector:
  print(data["traj_ids"])

>>> [0, 0, 0, ...]
>>> [0, 0, 0, ...]
...
```

This becomes an issue because I would like to insert episodes into the replay buffer __without__ padding. Instead, I would like to pad them when I sample them from the replay buffer, using `collate_fn`. However, there is a high probability that there are collisions between `traj_ids` when calling `buffer.sample(k)`. Then, `split_trajectories` will attempt to concatenate different episodes with identical `traj_ids`, which we do not want.

One solution could be to make `traj_ids` unique in the collector by keeping track of how many trajectories have been seen thus far.

**Comment by smorad** [comment]
Closing this as @vmoens has added a nice RDQN tutorial to the website!
